src/appleseed/renderer/modeling/camera/multilenscamera.cpp

```diff
@@ -0,0 +1,1328 @@
+
+            return true;
+        }
+
+        bool trace_ray_from_scene(Ray3d& ray, const int start_index = 0, const bool ignore_aperture_shape = false) const
+        {
+            double current_z;
+            if (start_index > 0)
+                current_z = get_z_offset(start_index);
+            else
+                current_z = 0;
+
+            for (auto lens_iter = m_lens_container.begin() + start_index; lens_iter != m_lens_container.end(); ++lens_iter)
+            {
+                LensElement current_element = *lens_iter;
+
+                double t = 0; // Parameter, at which ray intersects element
+                Vector3d normal(0, 0, 0); // Normal vector
+                if (current_element.radius == 0)
+                {
+                    // Aperture intersection
+                    t = (current_z - ray.m_org.z) / ray.m_dir.z;
+                }
+                else
+                {
+                    // Lens element intersection
+                    const Vector3d center = Vector3d(0, 0, current_z + current_element.radius); // Center of the lens element sphere
+                    if (!intersect_lens(ray, center, current_element.radius, t, normal))
+                        return false;
+                }
+
+                // Test intersection point and set as new origin if not out of bounds.
+                const Vector3d intersection = ray.point_at(t);
+                // Stop if intersection point lies outside of lens radius.
+                if (!is_inside(current_element, intersection, ignore_aperture_shape))
+                    return false;
+
+                ray.m_org = intersection;
+
+                if (current_element.radius != 0) {
+                    Vector3d t(0, 0, 0); // Refracted ray
+
+                    double ior_frac = 1;
+                    double prev_ior = 1;
+                    if (lens_iter != m_lens_container.begin())
+                    {
+                        auto prev_iter = std::prev(lens_iter);
+                        prev_ior = (!(*prev_iter).is_aperture) ? (*prev_iter).ior : 1;
+                    }
+                    ior_frac = prev_ior / current_element.ior;
+
+                    if (!refract(-ray.m_dir, normal, ior_frac, t))
+                        return false;
+                    ray.m_dir = t;
+                }
+
+                current_z += current_element.thickness;
+            }
+            return true;
+        }
+
+        bool trace_ray_from_film(Ray3d& ray, const int start_index = 0, const bool ignore_aperture_shape = false) const
+        {
+            double current_z;
+            if (start_index > 0)
+                current_z = get_z_offset(start_index);
+            else
+                current_z = m_total_z;
+
+            for (auto lens_iter = m_lens_container.rbegin() + start_index; lens_iter != m_lens_container.rend(); ++lens_iter)
+            {
+                LensElement current_element = *lens_iter;
```

Comment:
Const reference

Reply:
Done in 209c8188f88dbb0764ff2d20bf5c66a0df47b6f8